packages/orchestration/src/util/stream.ts

```diff
@@ -0,0 +1,351 @@
+  return {
+    index: choice.index,
+    message: {
+      role: 'assistant',
+      content: choice.delta.content,
+      tool_calls: transformStreamingToolCalls(choice.delta.tool_calls),
+      refusal: choice.delta.refusal
+    },
+    finish_reason: choice.finish_reason ?? '',
+    logprobs: choice.logprobs
+  };
+}
+
+function transformStreamingToolCalls(
+  toolCalls: ToolCallChunk[] | undefined
+): MessageToolCall[] | undefined {
+  if (!toolCalls || toolCalls.length === 0) {
+    return undefined;
+  }
+  return toolCalls?.map(toolCall => transformStreamingToolCall(toolCall));
+}
+
+function transformStreamingToolCall(toolCall: ToolCallChunk): MessageToolCall {
+  return {
+    index: toolCall.index,
+    id: toolCall.id ?? '',
+    type: toolCall.type ?? 'function',
+    function: {
+      name: toolCall.function?.name ?? '',
+      arguments: toolCall.function?.arguments ?? ''
+    }
+  };
+}
+
+/**
+ * @internal
+ */
+export function validateResponse(
+  response: OrchestrationStreamResponse<OrchestrationStreamChunkResponse>
+): void {
+  if (response._openStream) {
+    throw new Error(
+      "Stream wasn't closed properly. Please ensure the stream is closed after processing."
+    );
+  }
+
+  validateLlmModuleResult(response._data.module_results?.llm);
+
+  validateLlmModuleResult(response._data.orchestration_result);
+
+  validateChoices(response._data.module_results?.output_unmasking);
+}
+
+function validateLlmModuleResult(
+  llmModuleResult: Partial<LlmModuleResult> | undefined
+): void {
+  if (llmModuleResult) {
+    if (!llmModuleResult.usage) {
+      logger.warn('LlmModuleResult is missing usage information.');
+    }
+    if (!llmModuleResult.choices || llmModuleResult.choices.length === 0) {
+      logger.warn('LlmModuleResult must contain at least one choice.');
+    }
+
+    validateChoices(llmModuleResult.choices);
+  }
+}
+
+function validateChoices(choices: Partial<LlmChoice>[] | undefined): void {
+  if (choices) {
+    for (const choice of choices) {
+      if (!choice.message) {
+        logger.warn('LlmChoice is missing message information.');
+      } else {
+        validateMessage(choice.message);
+      }
+      if (!choice.finish_reason) {
+        logger.warn('LlmChoice is missing a finish reason.');
+      }
+      if (!choice.index && choice.index !== 0) {
+        logger.warn('LlmChoice must have a valid index.');
+      }
+    }
+  }
+}
+
+function validateMessage(message: Partial<ResponseChatMessage>): void {
+  if (!message.role) {
+    logger.warn('Message is missing role information.');
+  }
+  if (!message.content && !message.tool_calls) {
+    logger.warn('Message contains neither content nor tool calls.');
+  }
+
+  if (message.tool_calls) {
+    for (const toolCall of message.tool_calls) {
+      validateToolCall(toolCall);
+    }
+  }
+}
+
+function validateToolCall(toolCall: Partial<MessageToolCall>): void {
+  if (typeof toolCall.id !== 'string') {
```

Comment:
[pp] Also check if the `id` is not empty.

Reply:
AFAIK the id is never empty, the first chunk of a tool_call will always contain the id with the complete string, would you still do this additional check?